data/calculated_parser/functions.py

```diff
@@ -56,6 +56,40 @@ def magnitude(a, b):
     return np.sqrt(a ** 2 + b ** 2)
 
 
+def unstaggered_speed(u_vel, v_vel, x_dim_index, y_dim_index):
+    """Calculate the speed of seawater current from u and v velocity component
+    array that are on the u and v points of an Akawara-C staggered grid;
+    see https://en.wikipedia.org/wiki/Arakawa_grids
+
+    To correctly calculate the speed of the current, the velocity components have to be
+    "unstaggered" by interpolating their values to the T-grid points at the centres of the
+    grid cells. Here that is accomplished by averaging u(i-1) and u(i) values to get u values
+    at the T-points. Likewise, v(j-1) and v(j) values are averaged to get v values at the T-points.
+
+    With those arrays of unstaggered values, the speed of the current is calculated as the
+    element-wise magnitude of u and v:
+      np.sqrt(u ** 2 + v ** 2)
+    See: https://en.wikipedia.org/wiki/Hadamard_product_(matrices)
+
+    Paramters:
+    u_vel: ndarray
+    v_vel: ndarray
+    x_dim_index: float
+    y_dim_index: float
+
+    Returns:
+        ndarray -- speed of current
+    """
+    # The vel(t, z, y, x) arrays have been sliced to vel(t, y, x) by the time they get here,
+    # so we subtract 1 from the dimension index values passed in from the datasetconfig.json
+    # equation.
```

Comment:
Is this also true for transect plots? That's where we grab all depth levels and plot them against the drawn lines distance.
![image](https://user-images.githubusercontent.com/5572045/78247849-29b07c00-74c6-11ea-8bb7-f33670cbe7e2.png)

Reply:
Oh! Nice catch! I forgot to test Points, Lines, and Areas, but I just did so now, and they didn't error out. I will do a debugging trace to figure out how the dimension reduction is happening for the Lines case, and revise that comment to make it more general wrt to dimension reduction.

Reply:
Awesome. After that's done, I'll approve and merge.

Reply:
Arrgh! The velocity component arrays are `(t, depth, y, x)` when they get to `unstagger_speed()` for a transect. So, it is not doing the right thing at all: shifting `u` on the `gridY` axis, and `v` on the `depth` axis. :sob: 

Two options:
1. Assume that y and x are always the last 2 dimensions, and in that order, and pick them out of the `dims` attr using -2 and -1 indices.
2. Make `unstagger_speed()` specific to SalishSeaCast by hard-coding the dimension names as `gridX` and `gridY`. In that case, does the function name need to change to reflect its specificity?

Reply:
I think Option 1 is good. That's the layout I've been trying to implement.

Reply:
I would have a quick comment present saying the expected shape.

Reply:
It makes sense for us to push for `(t, depth, y, x)` or `(t, depth, lon, lat)` because that is the "standard" that NEMO uses, and NEMO is that "standard" for CONCEPTS, and CONCEPTS and Navigator are pretty closely linked...